File: abstractions/ruby/microsoft_kiota_abstractions/lib/microsoft_kiota_abstractions/serialization/serialization_writer_factory_registry.rb
require_relative 'serialization_writer_factory'

module MicrosoftKiotaAbstractions
  class SerializationWriterFactoryRegistry
    include SerializationWriterFactory

    class << self
      attr_accessor :default_instance
      def default_instance; @default_instance ||= SerializationWriterFactoryRegistry.new; end
    end

    def default_instance
      self.class.default_instance
    end

    def content_type_associated_factories
      @content_type_associated_factories ||= Hash.new
    end

    def get_serialization_writer(content_type)
      if !content_type
        raise Exception.new 'content type cannot be undefined or empty'
      end
      vendor_specific_content_type = content_type.split(';').first
      factory = @content_type_associated_factories[vendor_specific_content_type]
      if factory
        return factory.get_serialization_writer(vendor_specific_content_type)
      end

      clean_content_type = vendor_specific_content_type.gsub(/[^\/]+\+/i, '')
      factory = @content_type_associated_factories[clean_content_type]
      if factory
        return factory.get_serialization_writer(clean_content_type)
      end

      raise Exception.new "Content type #{contentType} does not have a factory to be serialized"
    end
    
  end
end

File: abstractions/swift/Source/MicrosoftKiotaAbstractions/Serialization/SerializationWriterFactoryRegistry.swift
public enum SerializationWriterFactoryRegistryErrors : Error {
    case registrySupportsMultipleTypesGetOneFactory
    case factoryNotFoundForContentType
}
public class SerializationWriterFactoryRegistry : SerializationWriterFactory {
    public var contentTypeAssociatedFactories = [String:SerializationWriterFactory]()
    public func getValidContentType() throws -> String {
        throw SerializationWriterFactoryRegistryErrors.registrySupportsMultipleTypesGetOneFactory
    }
    public func getSerializationWriter(contentType: String) throws -> SerializationWriter {
        guard let factory = contentTypeAssociatedFactories[contentType] else {
            throw SerializationWriterFactoryRegistryErrors.factoryNotFoundForContentType
        }
        return try factory.getSerializationWriter(contentType: contentType)
    }
}